src/base/path.cpp

```diff
@@ -0,0 +1,326 @@
+ * exception statement from your version.
+ */
+
+#include "path.h"
+
+#include <QDataStream>
+#include <QDir>
+#include <QFileInfo>
+#include <QMimeDatabase>
+#include <QRegularExpression>
+#include <QVector>
+
+#if defined(Q_OS_WIN)
+const Qt::CaseSensitivity CASE_SENSITIVITY = Qt::CaseInsensitive;
+#else
+const Qt::CaseSensitivity CASE_SENSITIVITY = Qt::CaseSensitive;
+#endif
+
+const int PATHLIST_TYPEID = qRegisterMetaType<PathList>();
+
+Path::Path(const QString &pathStr)
+    : m_pathStr {QDir::cleanPath(pathStr)}
+{
+}
+
+Path::Path(const std::string &pathStr)
+    : Path(QString::fromStdString(pathStr))
+{
+}
+
+Path::Path(const char pathStr[])
+    : Path(QString::fromLatin1(pathStr))
+{
+}
+
+bool Path::isValid() const
+{
+    if (isEmpty())
+        return false;
+
+#if defined(Q_OS_WIN)
+    const QRegularExpression regex {QLatin1String("[:?\"*<>|]")};
+#elif defined(Q_OS_MACOS)
+    const QRegularExpression regex {QLatin1String("[\\0:]")};
+#else
+    const QRegularExpression regex {QLatin1String("[\\0]")};
+#endif
+    return !m_pathStr.contains(regex);
+}
+
+bool Path::isEmpty() const
+{
+    return m_pathStr.isEmpty();
+}
+
+bool Path::isAbsolute() const
+{
+    return QDir::isAbsolutePath(m_pathStr);
+}
+
+bool Path::isRelative() const
+{
+    return QDir::isRelativePath(m_pathStr);
+}
+
+bool Path::exists() const
+{
+    return !isEmpty() && QFileInfo::exists(m_pathStr);
+}
+
+Path Path::rootItem() const
+{
+    const int slashIndex = m_pathStr.indexOf(QLatin1Char('/'));
+    if (slashIndex < 0)
+        return *this;
+
+    if (slashIndex == 0) // *nix absolute path
+        return createUnchecked(QLatin1String("/"));
+
+    return createUnchecked(m_pathStr.left(slashIndex));
+}
+
+Path Path::parentPath() const
+{
+    const int slashIndex = m_pathStr.lastIndexOf(QLatin1Char('/'));
+    if (slashIndex == -1)
+        return {};
+
+    if (slashIndex == 0) // *nix absolute path
+        return (m_pathStr.size() == 1) ? Path() : createUnchecked(QLatin1String("/"));
+
+    return createUnchecked(m_pathStr.left(slashIndex));
+}
+
+Path Path::filename() const
+{
+    const int slashIndex = m_pathStr.lastIndexOf('/');
+    if (slashIndex == -1)
+        return *this;
+
+    return createUnchecked(m_pathStr.mid(slashIndex + 1));
+}
+
+QString Path::extension() const
+{
+    const QString suffix = QMimeDatabase().suffixForFileName(m_pathStr);
+    if (!suffix.isEmpty())
+        return (QLatin1String(".") + suffix);
+
+    const int slashIndex = m_pathStr.lastIndexOf(QLatin1Char('/'));
+    const auto filename = QStringView(m_pathStr).mid(slashIndex + 1);
+    const int dotIndex = filename.lastIndexOf(QLatin1Char('.'));
+    return ((dotIndex == -1) ? QString() : filename.mid(dotIndex).toString());
+}
+
+bool Path::hasExtension(const QString &ext) const
+{
+    return (extension().compare(ext, Qt::CaseInsensitive) == 0);
+}
+
+bool Path::isBaseOf(const Path &other) const
```

Comment:
>> This method checks for any base path, i.e. if other is `/home/john/music/jazz` it will return true for `/home/john/music`, `/home/john`, `/home` and `/`.

I see. Then I would suggest `bool Path::hasAncestor(const Path &other)`.

>I would expect something like Path::hasParent(path) instead of other way around.

To elaborate a bit. IMO a class function should have `this` as the protagonist. It should provide usage like `Path("/main/a/b/c").hasAncestor(Path("/others"))`.